### PR TITLE
Add configurable time format

### PR DESCRIPTION
**Pull request by bmatherly**
This pull request is for review and staging until after the next release.

Add a new menu entry for time selection
image

This entry affects everything except for the rulers in the player and timeline
Here is a screenshot with some examples:
image

### Notes from the Design Discussion:

 **default to time clock format:**

Not done - currently defaults to SMPTE DF - which was the previous display. I can change it, but I am concerned that some people will be surprised when they upgrade Shotcut and the time display changes.

 **context menu on the time display:**

Not implemented. I made a Settings menu entry. I could also add a context menu on the time entry text field in the play if that is helpful. But people should be aware that it affects the entire application.

 **try to affect everywhere time is displayed:**

Done - with the exception of the Rulers. I can also change the rulers. But I notice that there is some "fudge factor" logic that we would have to overcome. I think that is probably to get around rounding errors.

**input parsing as well based on MLT’s interpretation of string to time:**

Done

### CODE REVIEW

**Comment by bmatherly** [comment]
@ddennedy  do you have a minute to review this? I would like to merge it. But I would like your comments since I deviated from the original design discussion somewhat.

**Comment by ddennedy** [comment]
No I’m traveling

**Comment by ddennedy** [comment]
> Add a new menu entry for time selection

I prefer this over only the context menu. However, I think "Drop Frame" sounds too scary on its own and both drop and non-drop should end with "Timecode."

> I can also change the rulers. 

Well, the rulers are special as they use timecode without frames. I think it is good to leave them as-is.

> currently defaults to SMPTE DF

I agree to not change the default. But mine defaulted to Non-Drop Frame.

File: src/mainwindow.ui [hunk 1]
```diff
@@ -202,6 +202,15 @@
      <addaction name="actionPreview540"/>
      <addaction name="actionPreview720"/>
     </widget>
+    <widget class="QMenu" name="menuTimeFormat">
+     <property name="title">
+      <string>Time Format</string>
+     </property>
+     <addaction name="actionTimeFrames"/>
+     <addaction name="actionTimeClock"/>
```
Comment (ddennedy) [inline src/mainwindow.ui:210]: Please move "Time Format" to just below "Timeline" in the Settings menu. And move "Frames" to the bottom of the sub-menu.

File: src/settings.cpp [hunk 2]
```diff
@@ -1278,3 +1278,14 @@ void ShotcutSettings::setBackupPeriod(int seconds)
 {
     settings.setValue("backupPeriod", seconds);
 }
+
+mlt_time_format ShotcutSettings::timeFormat() const
+{
+    return (mlt_time_format)settings.value("timeFormat", mlt_time_smpte_ndf).toInt();
```
Comment (ddennedy) [inline src/settings.cpp:1284]: Change default to `mlt_time_smpte_df`

**Review by ddennedy — CHANGES_REQUESTED** [review]

**Comment by bmatherly** [comment]
> However, I think "Drop Frame" sounds too scary on its own and both drop and non-drop should end with "Timecode."

Done. I put a different spin on it. Let me know what you think.

> I agree to not change the default. But mine defaulted to Non-Drop Frame.

Fixed

image

File: src/settings.cpp (hunk 2 above)
Comment (bmatherly) [reply src/settings.cpp]: Done

File: src/mainwindow.ui (hunk 1 above)
Comment (bmatherly) [reply src/mainwindow.ui]: Done

**Pull request merged**